File: Conlang.cs
/** 
File: Conlang.cs 
Date: 6/9/17
Author: @crosscripter
Description: A simple translator for my personal constructed language.
usage: $ conlang "<phrase|file>"

Copyright (C) @crosscripter 2017-2018.
All Rights Reserved (R).
*/

using System;
using System.IO;
using System.Linq;
using static System.Console;
using System.Collections.Generic;
using System.Text.RegularExpressions;


class Conlang
{
    static class Translator
    {
        static Dictionary<string,string> VowelMap = new Dictionary<string,string>
        {
            {"A", "I"},
            {"E", "O"},
            {"I", "A"},
            {"O", "E"},
            {"U", "Ey"}
        };

        static Dictionary<string,string> ConsonantMap = new Dictionary<string,string>
        {
            {"B", "Y"},
            {"D", "T"},
            {"F", "Ts"},
            {"G", "D"},
            {"H", "R"},
            {"J", "S"},
            {"K", "L"},
            {"L", "K"},
            {"M", "Z"},
            {"P", "G"},
            {"R", "V"},
            {"S", "M"},
            {"T", "Sh"},
            {"V", "F"},
            {"W", "'H"},
            {"Y", "'"},
            {"Z", "H"}
        };

        static bool IsVowel(string letter) => VowelMap.ContainsKey(letter.ToUpper());
        
        static bool IsConsonant(string letter) => ConsonantMap.ContainsKey(letter.ToUpper());

        enum Case { Upper, Lower }

        static Case GetCase(string letter) => letter.ToUpper() == letter ? Case.Upper : Case.Lower;

        static string SetCase(string text, Case @case)
        {
            switch (@case) 
            {
                case Case.Upper: return text.ToUpper();
                case Case.Lower: return text.ToLower();
                default: return text;
            }
        }

        static string ReplaceCase(string text, string pattern, string replacement)
        {
            var casing = GetCase(text);
            replacement = SetCase(replacement, casing);
            return Regex.Replace(text, pattern, replacement, RegexOptions.IgnoreCase);
        }

        public static string TranslateLetter(string letter)
        {
            var map = IsVowel(letter) ? VowelMap 
                    : IsConsonant(letter) ? ConsonantMap 
                    : null;

            if (map == null) return letter;
            var casing = GetCase(letter);
            var newLetter = map[letter.ToUpper()];
            return SetCase(newLetter, casing);
        }

        public static string TransformKSounds(string text) => ReplaceCase(text, "[CQX]", "K");

        public static string TranslateWord(string word)
        {   
            WriteLine("TranslateWord");
            word = TransformKSounds(word);
            var letters = word.ToCharArray();
            WriteLine($"TranslateWord: letters = {string.Join(",", letters)}");
            letters = letters.Select(l => TranslateLetter(l.ToString())).ToArray();
            return string.Join(string.Empty, letters);
        }

        public static string Translate(string text)
        {
            WriteLine("Translate");
            var words = text.Split(' ');
            WriteLine($"Translate: words = {string.Join(",", words)}");
            words = words.Select(TranslateWord).ToArray();
            WriteLine($"Translate: words translated = {string.Join(",", words)}");
            return string.Join(" ", words);
        }
    }

    static void Main(string[] args)
    {
        // Read the commandline arguments.
        // If we have none display usage and exit
        if (args.Length == 0)
        {
            WriteLine(@"Usage: conlang ""<phrase|file>""");
            return;
        }

        // Check the type of the argument passed
        var arg = args[0];

        // Argument is a file, text comes from file content
        var isFile = File.Exists(arg);
        var text = isFile ? File.ReadAllText(arg) : arg;

        // Translate text and display
        var translation = Translator.Translate(text);
        WriteLine($"translation: {translation}, from file? {isFile}");
    }
}

File: conlang.js
var fs = require('fs');

var vowels = ['A', 'E', 'I', 'O', 'U'];

function isVowel(letter) { 
	return vowels.indexOf(letter.toUpperCase()) !== -1; 
}

var consonants = [
	'B', 'C', 'D', 'F', 'G', 'H', 'J', 
	'K', 'L', 'M', 'N', 'O', 'P', 'Q', 
	'R', 'S', 'T', 'V', 'W', 'X', 'Y', 
	'Z'
];

function isConsonant(letter) {
	return !isVowel(letter);
}

var letterMap = {
	'A': 'I', 
	'I': 'A',
	'E': 'O', 
	'O': 'E',
	'U': 'E',
	'B': 'Y',
	'D': 'T',
	'F': 'TS',
	'G': 'D',
	'H': 'R',
	'J': 'S',
	'K': 'L',
	'L': 'K',
	'M': 'Z',
	'P': 'G',
	'Q': 'P',
	'R': 'V',
	'S': 'M',
	'T': 'SH',
	'V': 'F',
	'W': 'KH',
	'Y': "'A",
	'Z': 'H'
};

function getCase(letter) {
	return letter.toUpperCase() === letter ? "upper" : "lower";
}

function setCase(letter, casing) {
	return casing === "upper" ? letter.toUpperCase() : letter.toLowerCase();
}

function mapLetter(letter) {
	var originalLetter = letter;
	letter = letter.toUpperCase();

	switch (letter) {
		case 'C':
		case 'X':
			letter = 'K';
		default:
			var newLetter = letterMap[letter] || letter;
			break;
	}

	var casing = getCase(originalLetter);
	return setCase(newLetter, casing);
}

function doubleLetters(word) {
	return word.replace(/([^AEIOU]{1,2})\1/ig, function(s, $1) { 
		return s.substring(0, $1.length); 
	});
}

function replaceCase(source, letter) {
	var casing = getCase(source);
	letter = setCase(letter, casing);
	return letter;
}

function tripleConsonant(word) {
	return word.replace(/([^AEIOUHY]{3})/gi, function(s) {
		return s.substring(0, 2) + replaceCase(s, 'a') + s[s.length - 1];
	});
}

function endingConsonant(word) {
	if (!/(SH|TS|KH|NT)$/i.test(word)) {	
		word = word.replace(/([^AEIOU\W]{2})$/i, '$1' + replaceCase(word, 'a'));
	}
	return word;
}

function doubleO(word) {
	return word.replace(/OO/i, replaceCase(word, 'u'));
}

function doubleE(word) {
	return word.replace(/EE/i, replaceCase(word, "ey"));
}

function endingO(word) {
	return word.replace(/([A-Z]+)O$/i, '$1' + replaceCase(word, 'a'));
}

function mapWord(word) {
	var newWord = word.split('').map(mapLetter).join('');
	var transforms = [doubleLetters, tripleConsonant, endingConsonant, doubleO, doubleE, endingO];
	transforms.forEach(function(transform) { newWord = transform(newWord); });
	return newWord;
}

function translate(phrase) {
	return phrase.replace(/(\w+)\b/g, mapWord);
}

function random(items) {
	return items[Math.floor(Math.random() * items.length)];
}

var text = fs.readFileSync("C:/users/mikes/dropbox/docs/kjv.txt", "utf-8");
fs.writeFileSync("C:/users/mikes/dropbox/docs/kjv.conlang.txt", translate(text), "utf-8");
